File: packages/ios-im-ui/Sources/FlareIMUI/Models/FlareTimeFormat.swift
import Foundation

/// Locale-aware time labels for message and conversation rows — the rule every platform shares.
///
/// `messageTime` is the clock in the locale's convention: a two-digit hour in 24-hour locales
/// ("09:05"), the numeric hour with the day period in 12-hour ones ("9:05 AM").
/// `conversationTime` is that clock on the same calendar day, `yesterday` on the previous one,
/// the numeric month and day within the same year, and the numeric year, month and day before
/// it — each in the locale's order. Days and years are counted in the local time zone.
public enum FlareTimeFormat {
    public static func messageTime(_ date: Date, locale: Locale = .current) -> String {
        formatter(clockPattern(locale), locale).string(from: date)
    }

    public static func conversationTime(_ date: Date, yesterday: String, locale: Locale = .current,
                                        now: Date = Date()) -> String {
        let calendar = Calendar.current
        if calendar.isDate(date, inSameDayAs: now) { return messageTime(date, locale: locale) }
        if let previousDay = calendar.date(byAdding: .day, value: -1, to: now),
           calendar.isDate(date, inSameDayAs: previousDay) {
            return yesterday
        }
        let template = calendar.isDate(date, equalTo: now, toGranularity: .year) ? "Md" : "yMd"
        return formatter(pattern(template, locale), locale).string(from: date)
    }

    /// The label of a timeline date separator. Every bubble already shows its own time, so separators
    /// mark days only: `today` and `yesterday` (the host's words) for those days, then the locale's
    /// numeric month and day within the current year, and its year, month and day before it. Absolute,
    /// so a separator never goes stale while the conversation stays open.
    public static func timelineDateLabel(_ date: Date, today: String, yesterday: String,
                                         locale: Locale = .current, now: Date = Date()) -> String {
        let zone = TimeZone.current
        let day = localDay(date, zone)
        let currentDay = localDay(now, zone)
        if day == currentDay { return today }
        if day == currentDay - 1 { return yesterday }
        let calendar = Calendar.current
        let template = calendar.isDate(date, equalTo: now, toGranularity: .year) ? "Md" : "yMd"
        return formatter(pattern(template, locale), locale).string(from: date)
    }

    /// True when a timeline date separator belongs before a message sent at `currentMs`: the first
    /// dated message (`previousMs` is 0), or the first one on another local calendar day. Undated
    /// messages (0) never start a day. Decided on day numbers in `timeZone`, never on formatted text,
    /// so a timeline can ask for every row.
    public static func startsTimelineDay(previousMs: Int64, currentMs: Int64, timeZone: TimeZone = .current) -> Bool {
        guard currentMs > 0 else { return false }
        guard previousMs > 0 else { return true }
        return localDay(milliseconds: previousMs, timeZone) != localDay(milliseconds: currentMs, timeZone)
    }

    /// The local calendar day as a number: days since the epoch in `zone`, counted from local midnight.
    static func localDay(_ date: Date, _ zone: TimeZone) -> Int {
        let local = date.timeIntervalSince1970 + TimeInterval(zone.secondsFromGMT(for: date))
        return Int((local / 86_400).rounded(.down))
    }

    static func localDay(milliseconds: Int64, _ zone: TimeZone) -> Int {
        localDay(Date(timeIntervalSince1970: TimeInterval(milliseconds) / 1000), zone)
    }

    /// The locale's hour-and-minute pattern; in a 24-hour locale the hour is widened to two digits
    /// ("H:mm" → "HH:mm"). Quoted literals ("HH 'h' mm") are left as they are.
    static func clockPattern(_ locale: Locale) -> String {
        let pattern = DateFormatter.dateFormat(fromTemplate: "jm", options: 0, locale: locale) ?? "HH:mm"
        var fields = ""
        var quoted = false
        for character in pattern {
            if character == "'" { quoted.toggle() } else if !quoted { fields.append(character) }
        }
        // h / K count a 12-hour clock; a / b / B are its day periods.
        guard !fields.contains(where: { "hKabB".contains($0) }) else { return pattern }
        var widened = ""
        var previous: Character?
        quoted = false
        for character in pattern {
            if character == "'" { quoted.toggle() }
            let hourField = !quoted && (character == "H" || character == "k")
            if hourField {
                // Emit each run of the hour field once, two letters wide.
                if previous != character { widened.append(character); widened.append(character) }
            } else {
                widened.append(character)
            }
            previous = hourField ? character : nil
        }
        return widened
    }

    private static let cache = NSCache<NSString, DateFormatter>()
    private static let patterns = NSCache<NSString, NSString>()

    /// The locale's pattern for a date `template` ("Md", "yMd"), resolved once per locale.
    private static func pattern(_ template: String, _ locale: Locale) -> String {
        let key = "\(locale.identifier)|\(template)" as NSString
        if let cached = patterns.object(forKey: key) { return cached as String }
        let resolved = DateFormatter.dateFormat(fromTemplate: template, options: 0, locale: locale) ?? template
        patterns.setObject(resolved as NSString, forKey: key)
        return resolved
    }

    private static func formatter(_ pattern: String, _ locale: Locale) -> DateFormatter {
        let key = "\(locale.identifier)|\(pattern)" as NSString
        if let cached = cache.object(forKey: key) { return cached }
        let formatter = DateFormatter()
        formatter.locale = locale
        formatter.dateFormat = pattern
        cache.setObject(formatter, forKey: key)
        return formatter
    }
}

File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/FlareTimeFormat.kt
package com.flare.im.ui

import java.text.DateFormat
import java.text.SimpleDateFormat
import java.time.Instant
import java.time.ZoneId
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.ConcurrentHashMap

/**
 * Locale-aware time labels for host-formatted data (`FlareMessageData.timeLabel`,
 * `ConversationRowData.timestampLabel`) and the timeline's date separators, with the rules every
 * platform shares:
 *
 * - [messageTime]: hour and minute in the locale's convention — a two-digit 24-hour
 *   clock ("09:05") where the locale counts 24 hours, the locale's own 12-hour form
 *   ("9:05 AM", "上午9:05") where it does not.
 * - [conversationTime]: [messageTime] on the same calendar day, the host's
 *   `yesterday` label on the previous one, numeric month/day in the locale's order
 *   within the year, numeric year/month/day before it.
 * - [timelineDateLabel]: the same day buckets with the host's `today` word for the current day;
 *   [startsTimelineDay] decides where a separator goes.
 *
 * Calendar days are taken in the local time zone, from numeric fields; formatters are built once
 * per locale and shape, never per row.
 */
object FlareTimeFormat {
    fun messageTime(epochMs: Long, locale: Locale = Locale.getDefault()): String =
        formatIn(cachedFormat(locale, FlareDateShape.Clock), epochMs, ZoneId.systemDefault())

    fun conversationTime(
        epochMs: Long,
        yesterday: String,
        locale: Locale = Locale.getDefault(),
        nowMs: Long = System.currentTimeMillis(),
    ): String {
        val zone = ZoneId.systemDefault()
        val day = flareLocalEpochDay(epochMs, zone)
        val today = flareLocalEpochDay(nowMs, zone)
        return when (day) {
            today -> messageTime(epochMs, locale)
            today - 1 -> yesterday
            else -> dayLabel(epochMs, nowMs, locale, zone)
        }
    }

    /**
     * The label of a timeline date separator. Every bubble already shows its own time, so separators
     * mark days only: the host's [today] and [yesterday] words, then the locale's numeric month and day
     * in the current year, with the year when it differs.
     */
    fun timelineDateLabel(
        epochMs: Long,
        today: String,
        yesterday: String,
        locale: Locale = Locale.getDefault(),
        nowMs: Long = System.currentTimeMillis(),
        zone: ZoneId = ZoneId.systemDefault(),
    ): String {
        val day = flareLocalEpochDay(epochMs, zone)
        val current = flareLocalEpochDay(nowMs, zone)
        return when (day) {
            current -> today
            current - 1 -> yesterday
            else -> dayLabel(epochMs, nowMs, locale, zone)
        }
    }

    /**
     * True when a date separator belongs before the message sent at [currentMs]: the first dated
     * message ([previousMs] is 0) or one on another local calendar day than [previousMs]. An undated
     * message ([currentMs] is 0) gets none. Time gaps within a day never do.
     */
    fun startsTimelineDay(previousMs: Long, currentMs: Long, zone: ZoneId = ZoneId.systemDefault()): Boolean {
        if (currentMs <= 0) return false
        if (previousMs <= 0) return true
        return flareLocalEpochDay(previousMs, zone) != flareLocalEpochDay(currentMs, zone)
    }

    private fun dayLabel(epochMs: Long, nowMs: Long, locale: Locale, zone: ZoneId): String {
        val sameYear = Instant.ofEpochMilli(epochMs).atZone(zone).year == Instant.ofEpochMilli(nowMs).atZone(zone).year
        return formatIn(cachedFormat(locale, if (sameYear) FlareDateShape.MonthDay else FlareDateShape.FullDate), epochMs, zone)
    }

    private fun clockFormat(locale: Locale): DateFormat {
        val short = DateFormat.getTimeInstance(DateFormat.SHORT, locale)
        val pattern = (short as? SimpleDateFormat)?.toPattern().orEmpty()
        return when {
            !flareUsesTwelveHourClock(locale, pattern) -> SimpleDateFormat("HH:mm", locale)
            patternFields(pattern).any { it == 'h' || it == 'K' } -> short
            else -> SimpleDateFormat("h:mm a", locale)
        }
    }

    private enum class FlareDateShape { Clock, MonthDay, FullDate }

    private val formats = ConcurrentHashMap<Pair<Locale, FlareDateShape>, DateFormat>()

    private fun cachedFormat(locale: Locale, shape: FlareDateShape): DateFormat = formats.getOrPut(locale to shape) {
        when (shape) {
            FlareDateShape.Clock -> clockFormat(locale)
            FlareDateShape.MonthDay -> SimpleDateFormat(flareMonthDayPattern(shortDatePattern(locale)), locale)
            FlareDateShape.FullDate -> SimpleDateFormat(flareFullYearPattern(shortDatePattern(locale)), locale)
        }
    }

    private fun shortDatePattern(locale: Locale): String =
        (DateFormat.getDateInstance(DateFormat.SHORT, locale) as? SimpleDateFormat)?.toPattern() ?: "y/M/d"

    /** A [DateFormat] is not thread-safe and keeps its own time zone: format under its lock, in [zone]. */
    private fun formatIn(format: DateFormat, epochMs: Long, zone: ZoneId): String = synchronized(format) {
        format.timeZone = TimeZone.getTimeZone(zone)
        format.format(Date(epochMs))
    }
}

/** The local calendar day of [epochMs] in [zone], as days since the epoch: numeric, no formatter. */
internal fun flareLocalEpochDay(epochMs: Long, zone: ZoneId): Long {
    val offsetMs = zone.rules.getOffset(Instant.ofEpochMilli(epochMs)).totalSeconds * 1000L
    return Math.floorDiv(epochMs + offsetMs, 86_400_000L)
}

/**
 * Regions whose preferred hour cycle is 12-hour (CLDR supplemental `timeData`, the data
 * behind the `j` skeleton that ICU, Foundation and `Intl` resolve). The platform's short
 * time pattern is not used for this: JDK and older ICU builds disagree with CLDR for
 * regions such as CN (`ah:mm`) and MX (`HH:mm`).
 */
private val twelveHourRegions = setOf(
    "AC", "AE", "AG", "AL", "AQ", "AR", "AS", "AU", "BB", "BD", "BH", "BM", "BN", "BO", "BQ", "BS", "BT",
    "CA", "CC", "CL", "CO", "CP", "CQ", "CR", "CU", "CW", "CY", "DJ", "DM", "DO", "DZ", "EC", "EG", "EH",
    "ER", "ET", "FJ", "FM", "GD", "GH", "GM", "GR", "GT", "GU", "GY", "HK", "HM", "HN", "HT", "IN", "IQ",
    "JM", "JO", "KH", "KI", "KN", "KP", "KR", "KW", "KY", "LB", "LC", "LR", "LS", "LY", "MH", "MO", "MP",
    "MR", "MV", "MW", "MX", "MY", "NA", "NI", "NZ", "OM", "PA", "PE", "PG", "PH", "PK", "PR", "PS", "PW",
    "PY", "QA", "SA", "SB", "SD", "SG", "SL", "SO", "SS", "SV", "SY", "SZ", "TA", "TC", "TD", "TK", "TN",
    "TO", "TT", "TV", "TW", "UM", "US", "UY", "VC", "VE", "VG", "VI", "VU", "WS", "YE", "ZM",
)

/** Language-specific `timeData` entries that keep a 24-hour clock inside a 12-hour region. */
private val twentyFourHourLocales = setOf("fr-CA", "fr-HT", "nl-BQ", "nl-CW", "en-CC", "en-MV", "en-TK", "en-TV")

/**
 * Whether [locale] reads time on a 12-hour clock. Without a region the locale's own
 * [shortTimePattern] decides.
 */
internal fun flareUsesTwelveHourClock(locale: Locale, shortTimePattern: String): Boolean {
    val region = locale.country.uppercase(Locale.ROOT)
    return when {
        region.isEmpty() -> patternFields(shortTimePattern).any { it == 'h' || it == 'K' }
        "${locale.language}-$region" in twentyFourHourLocales -> false
        else -> region in twelveHourRegions
    }
}

/** A short date pattern without its year: the year field goes with the separator that joined it. */
internal fun flareMonthDayPattern(shortDatePattern: String): String {
    val parts = patternParts(shortDatePattern)
    val year = parts.indexOfFirst { it.field == 'y' }
    if (year < 0) return shortDatePattern
    val fieldBefore = parts.subList(0, year).any { it.field != null }
    val separator = if (fieldBefore) year - 1 else year + 1
    val drop = setOfNotNull(year, separator.takeIf { parts.getOrNull(it)?.field == null && it in parts.indices })
    return parts.filterIndexed { index, _ -> index !in drop }.joinToString("") { it.text }
}

/** A short date pattern with a full year ("M/d/yy" → "M/d/y"). */
internal fun flareFullYearPattern(shortDatePattern: String): String =
    patternParts(shortDatePattern).joinToString("") { if (it.field == 'y') "y" else it.text }

/** One run of a date pattern: a field (repeated letter) or literal text, quotes kept as written. */
private data class PatternPart(val field: Char?, val text: String)

private fun patternParts(pattern: String): List<PatternPart> {
    val parts = mutableListOf<PatternPart>()
    val literal = StringBuilder()
    fun flushLiteral() { if (literal.isNotEmpty()) { parts += PatternPart(null, literal.toString()); literal.clear() } }
    var i = 0
    while (i < pattern.length) {
        val c = pattern[i]
        when {
            c == '\'' -> {
                var j = i + 1
                while (j < pattern.length) {
                    if (pattern[j] != '\'') j++
                    else if (j + 1 < pattern.length && pattern[j + 1] == '\'') j += 2
                    else break
                }
                val end = minOf(j + 1, pattern.length)
                literal.append(pattern, i, end)
                i = end
            }
            c in 'a'..'z' || c in 'A'..'Z' -> {
                flushLiteral()
                var j = i
                while (j < pattern.length && pattern[j] == c) j++
                parts += PatternPart(c, pattern.substring(i, j))
                i = j
            }
            else -> { literal.append(c); i++ }
        }
    }
    flushLiteral()
    return parts
}

private fun patternFields(pattern: String): List<Char> = patternParts(pattern).mapNotNull { it.field }
